In [6]:
pdf_path = "test/pdf/dc/form/Blue_Book_2023_web.pdf"

# PDF Plumber

In [7]:
import pdfplumber
import pandas as pd
import numpy as np

pdf = pdfplumber.open(pdf_path)

In [8]:
print(pdf.pages[6].extract_words())

[{'text': 'Leaf', 'x0': 57.0, 'x1': 112.392, 'top': 71.92240000000004, 'doctop': 4607.9224, 'bottom': 95.92240000000004, 'upright': True, 'direction': 1}, {'text': 'stages', 'x0': 120.83999999999999, 'x1': 207.21599999999998, 'top': 71.92240000000004, 'doctop': 4607.9224, 'bottom': 95.92240000000004, 'upright': True, 'direction': 1}, {'text': '–', 'x0': 215.66400000000002, 'x1': 227.66400000000002, 'top': 71.92240000000004, 'doctop': 4607.9224, 'bottom': 95.92240000000004, 'upright': True, 'direction': 1}, {'text': 'cereals', 'x0': 236.11200000000002, 'x1': 330.864, 'top': 71.92240000000004, 'doctop': 4607.9224, 'bottom': 95.92240000000004, 'upright': True, 'direction': 1}, {'text': 'and', 'x0': 339.312, 'x1': 386.47200000000004, 'top': 71.92240000000004, 'doctop': 4607.9224, 'bottom': 95.92240000000004, 'upright': True, 'direction': 1}, {'text': 'grasses', 'x0': 394.9200000000001, 'x1': 496.44000000000005, 'top': 71.92240000000004, 'doctop': 4607.9224, 'bottom': 95.92240000000004, 'up

In [9]:
# print table
for table in pdf.pages[6].extract_tables():
    for row in table:
        print(row)

['One leaf stage\nEarly Late', None, 'Two leaf stage 2\n1 1\n2\nEarly Late', None]
['Three leaf stage\n3\n2\n2\n1\n1\n3\nEarly Late', None, 'Four leaf stage\n4\n3\n3\n2\n2\n4\n1\n1\nEarly Late', None]
['Five leaf stage\nWith one tiller\n4\n5\n3\nTiller\n2\n1', 'Six leaf stage\nWith two tillers\n5 6\n4\n3\nTiller\nTiller\n1 2', None, 'Cereal - early boot stage\nFlag leaf \n(the last leaf formed \nbefore the head \nemerges)\nSwelling caused \nby developing \nhead\nTiller']


In [15]:
pdf.pages[6].to_dict()

{'page_number': 7,
 'initial_doctop': 4536.0,
 'rotation': 0,
 'cropbox': [0.0, 0.0, 585.0, 756.0],
 'mediabox': [0.0, 0.0, 585.0, 756.0],
 'bbox': (0.0, 0.0, 585.0, 756.0),
 'width': 585.0,
 'height': 756.0,
 'rects': [{'x0': 529.0,
   'y0': 0.0,
   'x1': 585.0,
   'y1': 38.0,
   'width': 56.0,
   'height': 38.0,
   'pts': [(529.0, 0.0), (585.0, 0.0), (585.0, 38.0), (529.0, 38.0)],
   'linewidth': 0,
   'stroke': False,
   'fill': True,
   'evenodd': False,
   'stroking_color': None,
   'non_stroking_color': [0, 0, 0],
   'object_type': 'rect',
   'page_number': 7,
   'top': 718.0,
   'bottom': 756.0,
   'doctop': 5254.0},
  {'x0': 314.542,
   'y0': 261.286,
   'x1': 497.15999999999997,
   'y1': 421.21299999999997,
   'width': 182.618,
   'height': 159.92699999999996,
   'pts': [(314.542, 261.286),
    (497.15999999999997, 261.286),
    (497.15999999999997, 421.21299999999997),
    (314.542, 421.21299999999997)],
   'linewidth': 0,
   'stroke': False,
   'fill': True,
   'evenodd': Fa

In [10]:
pdf.pages[6].extract_text()

'Leaf stages – cereals and grasses\nOne leaf stage Two leaf stage 2\n1 1\n2\nEarly Late Early Late\nThree leaf stage Four leaf stage\n4\n3\n2\n3\n3\n2\n2\n2\n1\n1\n4\n3\n1\n1\nEarly Late Early Late\nFive leaf stage Six leaf stage Cereal - early boot stage\nWith one tiller With two tillers Flag leaf \n(the last leaf formed \n5 6 before the head \nemerges)\n4\n4\n5 3 estry\nor\nF\nd \nn\nSwelling caused  a\n3 e \nby developing  ur\nhead ult\nTiller gric\nTiller Tiller a A\n2 ert\n1 1 2 Tiller e: Alb\nc\nur\no\nS\nback to \nTable of Contents 3'

In [32]:
# print table
for table in pdf.pages[7].extract_tables():
    for row in table:
        print(row)

['Alternate leaves\nCotyledon stage', '1\nOne leaf stage']
['1\n2\nTwo leaf stage', '3\n2\n1\nThree leaf stage']
['Opposite leaves\nCotyledon stage', '2\n1\nTwo leaf stage']
['3\n4\n2\n1\nFour leaf stage', '6\n5\n3\n4\n2\n1\nSix leaf stage']
['Whorled leaves\nCotyledon stage', '1\nOne whorl stage']
['2\n1\nTwo whorl stage', '3\n2\n1\nThree whorl stage']
['Leaf stages – legume crops']
['Alfalfa and clovers\nFirst true leaf\nFirst trifoliate leaf\nSecond trifoliate leaf\nCotyledon\nSecond trifoliate leaf stage']


In [33]:
# print table
for table in pdf.pages[20].extract_tables():
    for row in table:
        print(row)

# print text
print(pdf.pages[20].extract_text())

Table 1
Example  Remaining  Clean water  Rinse batches  Total clean  Dilution factor  Single rinse 
  volume (A)  volume (B)  (C)  water  ((A&B)/A)C  volume equivalent
1  10  150  1  150  16  150
2  10  1200  1  1200  121  1200
3  10  50  3  150  216  2150
4  10  30  5  150  1024  10230
5  20  150  1  150  9  150
6  20  50  3  150  43  860
7  50  150  1  150  4  150
8  50  50  3  150  8  350
6.  Once dilution of the tank remainder is complete,  There are 3 main methods for sprayer tank clean-out. 
rinse out boom ends and any other plumping  Selection of the appropriate method depends on the 
components that were not open during dilution.  pesticide used. 
Inspect and clean filters.
Method A
7.  Consider modifications to your sprayer to improve 
Drain contents of tank followed by 1 or 2 water rinses, 
cleaning performance. A continuous rinse system 
2 ammonia rinses (NOT containing chlorine) 
uses a second pump to rinse the tank while it is 
followed by 2 water rinses (one just prior to

In [34]:
# print table
for table in pdf.pages[30].extract_tables():
    for row in table:
        print(row)

['Municipality', 'Site', 'Legal land location', 'Contact']
['Acadia, MD of', 'Acadia Valley Transfer Station', 'NW7-25-1-W4', '403-972-3879']
['Athabasca, County of', 'Colinton Waste Transfer Site', 'NE7-65-22-W4', '780-675-1117']
['Barrhead, County of', 'Barrhead Regional Landfill', 'SW3-60-4-W5', '780-674-3301']
['Beaver County', 'Beaver Regional Landfill', 'NE10-50-17-W4', '780-663-3888']
['Beaver County', 'Viking Transfer Station', 'NE31-47-12-W4', '780-718-7965']
['Big Lakes, MD of', 'High Prairie Regional Landfill', 'NW9-73-15-W5', '780-523-5955']
['Birch Hills County', 'Wanham Transfer Station', 'SW3-77-2-W6', '780-694-3793']
['Birch Hills County', 'Tangent Landfill', 'NE36-78-25-W5', '780-694-3793']
['Bonnyville, MD of', 'Goodridge Landfill', 'SW4-63-9-W4', '780-826-3951']
['Brazeau County', 'Drayton Valley', 'SE20-49-7-W5', '780-542-7777']
['Calgary, City of', 'Calgary – Forest Lawn Landfill  \n(includes MD #44)', '3801, 68 St. SE Calgary', '403-268-9947']
['Camrose, County of

In [35]:
# Turn tables into dataframes

dfs = []

for table in pdf.pages[31].extract_tables():
    df = pd.DataFrame(table[1:], columns=table[0])
    dfs.append(df)


In [36]:
dfs[0]

Municipality                                        Site  \
0               Lacombe County             Eckville Waste Transfer Station   
1               Lacombe County          Mirror-Alix Waste Transfer Station   
2               Lacombe County             Prentiss Waste Transfer Station   
3                Lamont County                    Lamont Regional Landfill   
4                 Leduc County               Nisku Sewage Transfer Station   
5                 Leduc County                       Thorsby – County Yard   
6    Lesser Slave River, MD of             Flatbush Waste Transfer Station   
7        Lethbridge, County of             Coaldale Waste Transfer Station   
8        Lethbridge, County of         Iron Springs Waste Transfer Station   
9        Lethbridge, County of            Nobleford Waste Transfer Station   
10       Lethbridge, County of        Picture Butte Waste Transfer Station   
11            MacKenzie, MD of                High Level Regional Landfill   
12          Minburn, County of                         Vegreville Landfill   
13          Minburn, County of                          Mannville Landfill   
14          Minburn, County of                           Ranfurly Landfill   
15        Mountain View County                     Didsbury – near airport   
16            Newell County of                    Newell Regional Landfill   
17           Newell, County of              Bassano Waste Transfer Station   
18      Northern Lights, MD of    Dixonville – Long Lake Regional Landfill   
19  Northern Lights, County of                 North Star Transfer Station   
20  Northern Lights, County of                 Hawk Hills Transfer Station   
21     Northern Sunrise County                Nampa Waste Transfer Station   
22     Northern Sunrise County                Peace Regional Landfill Site   
23       Paintearth, County of                     Castor Transfer Station   
24       Paintearth, County of                 Coronation Transfer Station   
25             Parkland County                        Stony Plain Landfill   
26             Parkland County                   Tomahawk Transfer Station   
27                Peace, MD of                      Griffin Creek Landfill   
28        Pincher Creek, MD of                    Cowley Regional Landfill   
29        Pincher Creek, MD of                             Pincher Station   
30               Ponoka County                           Bluffton Landfill   
31               Ponoka County               Ponoka Waste Transfer Station   
32              Provost, MD of                   Provost Regional Landfill   
33             Red Deer County            Horn Hill Waste Transfer Station   
34           Rocky View, MD of                   Bragg Creek Transfer Site   
35           Rocky View, MD of                    Langdon Transfer Station   
36           Rocky View, MD of                      Irricana Transfer Site   
37         Saddle Hills County                  Blueberry Transfer Station   
38           Smoky Lake County                 Smoky Lake Transfer Station   
39          Smoky River, MD of    Smoky River Waste  \nCommission Landfill   
40             Special Area #2             Bindloss Waste Transfer Station   
41             Special Area #2                Hanna Waste Transfer Station   
42             Special Area #3                 Oyen Waste Transfer Station   
43             Special Area #3  Youngstown – Big Country Regional Landfill   

   Legal land location                  Contact  
0         NW34-39-3 W5             403-782-6601  
1        NW24-40-23-W4             403-782-6601  
2         NW7-40-25-W4             403-782-6601  
3         NW7-56-18-W4             780-895-2585  
4        SW31-50-24-W4             780-955-3555  
5         SE17-49-1-W5             780-955-3555  
6         NW36-65-2-W5             780-681-3929  
7         SW23-9-20-W4             403-328-5525  
8        SW27-11-20-W4             403-328-5525  
9      

In [37]:
# print table
for table in pdf.pages[32].extract_tables():
    for row in table:
        print(row)


['Municipality', 'Site', 'Legal land location', 'Contact']
['Special Area #4', 'Monitor Waste Transfer Station', 'NW32-34-4-W4', '403-577-3523']
['Spirit River, MD of', 'Spirit River Landfill', 'SW31-77-5-W6', '780-864-3500']
['St. Paul, County of', 'Mallaig Waste Transfer Station', 'NE24 60 10 W4', '780-645-3301']
['St. Paul, County of', 'Evergreen Regional Landfill', 'NW16-56-10-W4', '780-645-3301']
['Starland County', 'Rumsey Waste Transfer Station', 'SW24-33-21 W4', '403-772-3793']
['Starland County', 'Michichi Waste Transfer Station', 'NW19 30 18 W4', '403-772-3793']
['Stettler, County of', 'Stettler Regional Landfill', 'NW22 40 19 W4', '403-742-4441']
['Strathcona County', 'Fort Saskatchewan Transfer Station  \nand Recycling Depot', '11121 88 Ave. Ft. Sask.', '780-417-7134']
['Sturgeon County', 'Roseridge Landfill', 'SW36 55 25 W4', '780-939-8325']
['Taber, MD of', 'Enchant Waste Transfer Station', 'NW16 14 18 W4', '403-223-8735']
['Taber, MD of', 'Grassy Lake Waste Transfer Stat

In [38]:
# Turn tables into dataframes

dfs = []

for table in pdf.pages[32].extract_tables():
    df = pd.DataFrame(table[1:], columns=table[0])
    dfs.append(df)


In [39]:
dfs[0]

Municipality  \
0                 Special Area #4   
1             Spirit River, MD of   
2             St. Paul, County of   
3             St. Paul, County of   
4                 Starland County   
5                 Starland County   
6             Stettler, County of   
7               Strathcona County   
8                 Sturgeon County   
9                    Taber, MD of   
10                   Taber, MD of   
11                   Taber, MD of   
12                   Taber, MD of   
13            Thorhild, County of   
14           Two Hills, County of   
15           Two Hills, County of   
16     Vermilion River, County of   
17     Vermilion River, County of   
18     Vermilion River, County of   
19                  Vulcan County   
20                  Vulcan County   
21                  Vulcan County   
22                  Vulcan County   
23                  Vulcan County   
24              Wainwright, MD of   
25               Wainwright MD of   
26               Wainwright MD of   
27               Wainwright MD of   
28              Warner, County of   
29              Warner, County of   
30                Westlock County   
31          Wetaskiwin, County of   
32               Wheatland County   
33               Wheatland County   
34               Wheatland County   
35            Willow Creek, MD of   
36               Woodlands County   
37  Woodlands County/\nWhitecourt   
38              Yellowhead County   

                                                 Site  \
0                      Monitor Waste Transfer Station   
1                               Spirit River Landfill   
2                      Mallaig Waste Transfer Station   
3                         Evergreen Regional Landfill   
4                       Rumsey Waste Transfer Station   
5                     Michichi Waste Transfer Station   
6                          Stettler Regional Landfill   
7   Fort Saskatchewan Transfer Station  \nand Recy...   
8                                  Roseridge Landfill   
9                      Enchant Waste Transfer Station   
10                 Grassy Lake Waste Transfer Station   
11                       Taber Waste Transfer Station   
12                          Vauxhall Transfer Station   
13                          Thorhild Transfer Station   
14                        Two Hills Regional Landfill   
15                     Willingdon Seed Cleaning Plant   
16                    Marwayne Waste Transfer Station   
17             Paradise Valley Waste Transfer Station   
18                         Vermilion Transfer Station   
19                         Mossleigh Transfer Station   
20                      Vulcan Waste Transfer Station   
21                            Lomond Transfer Station   
22                              Milo Transfer Station   
23                          Champion Transfer Station   
24                Wainwright – Parrish and Heimbecker   
25                                     Irma Shop Yard   
26  Edgerton (south of the railway tracks on  \nRa...   
27                                   Chauvin Landfill   
28                         Border Seed Cleaning Plant   
29                       Sunshine Seed Cleaning Plant   
30                         Westlock Regional Landfill   
31                       Peace Hills Transfer Station   
32                         Hussar Waste Transfer Site   
33                       Standard Waste Transfer Site   
34                 Wheatland West Waste Transfer Site   
35                     Willow Creek Regional Landfill   
36                     Fort Assiniboine – County Yard   
37                       Whitecourt Regional Landfill   
38                   Parkcourt Waste Transfer Station   

        Legal land location       Contact  
0              NW32-34-4-W4  403-577-3523  
1              SW31-77-5-W6  780-864-3500  
2             NE24 60 10 W4  780-645-3301  
3             NW16-56-10-W4  780-645-3301  
4             SW24-33-21 W4  403-772-37

In [40]:
# print table
for table in pdf.pages[48].extract_tables():
    for row in table:
        print(row)


['Group 2', '', 'ALS/AHAS inhibitors. These chemicals block the normal function of an enzyme called acetolactate (ALS) actohydroxy acid \n(AHAS). This enzyme is essential in amino acid (protein) synthesis. Without proteins, plants starve to death.', None, None, None, None, None]
['Imidazolinones', '', 'AC 299, 263 120 AS', '', 'Altitude FX3', '', '', '']
['', '', 'imazamox', '', 'Ares SN\nAltitude FX3\nAmity WDG\nBenz\nDavai', 'Dakota\nDavai A Plus\nIPCO Radian\nMizuna\nSamurai', 'Samurai Master\nNext 70 WDG\nPython\nQuasar', 'Solo/Solo ADV\nSolo Ultra Q\nVenim\nViper']
['', '', 'imazamox + imazethapyr', '', 'Davai\nDuet', 'Judo\nNinja', 'Ninja Master\nOdyssey/Odyssey NXT', 'Odyssey Ultra Q']
['', '', 'imazapyr', '', 'Arsenal', 'Ares SN', 'Quasar', '']
['', '', 'imazethapyr', '', 'Kamikaze', 'Phantom', 'Pursuit', '']
['Sulfonylamino-\ncarbonyltriazolinones', None, 'flucarbazone sodium', '', 'Denali CM\nEverest 3.0 AG', 'Inferno Duo\nInferno Trio', 'Himalaya\nHimalaya Extra', 'Himalaya 

In [41]:
# Turn tables into dataframes

dfs = []

for table in pdf.pages[48].extract_tables():
    df = pd.DataFrame(table[1:], columns=table[0])
    dfs.append(df)


In [42]:
dfs[0]

Group 2        \
0                          Imidazolinones         
1                                                 
2                                                 
3                                                 
4                                                 
5   Sulfonylamino-\ncarbonyltriazolinones  None   
6                                    None  None   
7                           Sulfonylureas         
8                                                 
9                                                 
10                                                
11                                                
12                                                
13                                                
14                                                
15                                                
16                               Pyrazole         
17                     Triazolpyramidines         
18                                                
19                            Triazolones         

   ALS/AHAS inhibitors. These chemicals block the normal function of an enzyme called acetolactate (ALS) actohydroxy acid \n(AHAS). This enzyme is essential in amino acid (protein) synthesis. Without proteins, plants starve to death.  \
0                                  AC 299, 263 120 AS                                                                                                                                                                                       
1                                            imazamox                                                                                                                                                                                       
2                              imazamox + imazethapyr                                                                                                                                                                                       
3                                            imazapyr                                                                                                                                                                                       
4                                         imazethapyr                                                                                                                                                                                       
5                                 flucarbazone sodium                                                                                                                                                                                       
6                             propoxycarbazone sodium                                                                                                                                                                                       
7                                       chlorsulfuron                                                                                                                                                                                       
8                              ethametsulfuron methyl                                                                                                                                                                                       
9                                        halosulfuron                                                                                                                                                                                       
10                                 metsulfuron-methyl                                                                                                                                                                                       
11                                       nicosulfuron                                                                 

Merged cells are just `None`

In [43]:
dfs[1]

Herbicide group classification by mode of action None  \
0                                  Chemical family        
1                                          Group 3        
2                                  Dinitroanilines        

                                                None  None       None  \
0                                 Active ingredients        Found in*   
1  Microtubule assembly inhibitors. These chemica...  None       None   
2                                      ethalfluralin             Edge   

                        None  None  None  
0                       None  None  None  
1                       None  None  None  
2  Advantage \nEthalfluralin

Tables can be splitted into two pages, and PDFPlumber can't handle it.

In [44]:
# Turn tables into dataframes

dfs = []

for table in pdf.pages[49].extract_tables():
    df = pd.DataFrame(table)
    dfs.append(df)


In [45]:
dfs[0]

0     1                                                  2  \
0                                                                 trifluralin   
1            Group 4        Synthetic auxins. These chemicals disrupt plan...   
2      Benzoic acids                                                  dicamba   
3   Carboxylic acids  None                                aminocyclopyrachlor   
4               None  None                                         fluroxypyr   
5               None  None                                         quinclorac   
6     Picolinic acid  None                                       aminopyralid   
7               None  None                                         clopyralid   
8               None  None                                         halauxifen   
9               None  None                                           picloram   
10           Phenoxy                                                    2,4-D   
11              None                                     dichlorprop (2,4-DP)   
12              None                                                   2,4-DB   
13              None                                                     MCPA   
14              None                                                     MCPB   
15              None                                          mecoprop (MCPP)   

       3                                                  4  \
0         Advantage Trifluralin\nAdvantage Trillaite \n1...   
1   None                                               None   
2           Ammo\nAmmo Extra\nAviator ll\nDiflux\nDisha 480   
3                                               Navius FLEX   
4         Akito\nAltitude FX3\nAudible\nAxial Xtreme\nAx...   
5                                                    Clever   
6                                                Reclaim II   
7         Advantage Clopyralid\nAkito\nCT Mix 360\nCo-op...   
8                    Cirpreme\nCo-op Exhilarate\nExhilarate   
9                                                 Grazon XC   
10        2,4-D\nCo-op Convex\nEnforcer D\nEstaprop XT\n...   
11                                                Desormone   
12                                                   2,4-DB   
13        Badge\nBrilliant\nBroadside\nBuctril M\nCanuck...   
14                                            Clovitox Plus   
15                                               Mecoprop-P   

                                                    5  \
0                 Bonanza\nFortress MicroActiv\nRival   
1                                                None   
2   Distinct\nEngenia\nExpress FX\nIntruvix Herbic...   
3                                            TruRange   
4   Enforcer M\nErebus Xtreme\nEsteem\nFlurox-24\n...   
5                                             Facet L   
6                                          Restore II   
7   Denali CM\nEsteem\nFBN Clopyralid\nInertia\nIP...   
8              IPCO Exhilarate\nParadigm PRE\nPixxaro   
9                                          Tordon 22K   
10  Grazon\nIPCO Convex\nIPCO Dichlorprop DX\nIPCO...   
11                                        Estaprop XT   
12                                        Caliber 625   
13  Enforcer M\nEsteem\nExhilarate\nForceFighter M...   
14                                      Tropotox Plus   
15                                              Sword   

                                                    6  \
0                                     Thrill\nTreflan   
1                                                None   
2        Oracle\nPilot\nPulsar\nRevenge A\nRevenge AE   
3                                                       
4   Foxxy CM\nFoxxy CRX\nFoxxy M\nFoxxy Pro\nFoxxy...   
5                                           Ingenious   
6                                                       
7   Momentum\nClobber\nClobber M\nFoxxy CM\nFoxxy ...   
8                                  Prospect\nProminex   
9                            

In [46]:
# Turn tables into dataframes

dfs = []

for table in pdf.pages[51].extract_tables():
    df = pd.DataFrame(table[1:], columns=table[0])
    dfs.append(df)


In [47]:
dfs[1]

Product                     Company  \
0            2,4-D 600 amine (PCP# 17511)                        IPCO   
1            2,4-D 600 amine (PCP# 14726)          Nufarm Agriculture   
2             2,4-D 600 amine (PCP# 5931)  Loveland Products \nCanada   
3            2,4-D 600 amine (PCP# 29248)                        IPCO   
4            2,4-D 700 ester (PCP# 27819)                        IPCO   
5            2,4-D 700 ester (PCP# 27820)          Nufarm Agriculture   
6  MPower 2,4-D 700 ester  \n(PCP# 30460)                 NewAgco Inc   
7      2,4-D 700 ester (PCP# 27818) Salvo  Loveland Products \nCanada   

  Active ingredient                 Formulation      Container size  
0           560 g/L                    Solution         10 L, 115 L  
1           564 g/L                    Solution                10 L  
2           560 g/L                    Solution                10 L  
3           560 g/L                    Solution                10 L  
4           660 g/L  Emulsifiable \nconcentrate  10 L, 115 L, 450 L  
5           660 g/L  Emulsifiable \nconcentrate  10 L, 115 L, 500 L  
6           660 g/L  Emulsifiable \nconcentrate  10 L, 115 L, 510 L  
7           660 g/L  Emulsifiable \nconcentrate         10 L, 115 L

In [55]:
# Turn tables into dataframes

dfs = []

for table in pdf.pages[54].extract_tables():
    df = pd.DataFrame(table[1:], columns=table[0])
    dfs.append(df)


In [57]:
dfs[1]

Timing  \
0             Pre-plant incorporated   
1                     Pre-emergence*   
2  Post-emergence (1 - 6 leaf stage)   

                                                Rate  \
0  Light and sand: 0. 85 - 1.25 L/acre\nLoam and ...   
1                                               None   
2                                               None   

                                  Crop  
0  Silage corn\nField corn\nSweet corn  
1                                 None  
2                                 None

In [60]:
# Turn tables into dataframes

dfs = []

for table in pdf.pages[80].extract_tables():
    df = pd.DataFrame(table[1:], columns=table[0])
    dfs.append(df)


In [61]:
dfs[1]

Treatment                Rate (per acre)  \
0                  None                      Soil type   
1                  None  Medium texture  \n(1 - 3% OM)   
2  Early season control                         162 mL   
3      Extended control                         202 mL   

                                 None                Treated per jug  \
0                                None                      Soil type   
1  Medium-fine texture  \n(3 - 6% OM)  Medium texture  \n(1 - 3% OM)   
2                                None                    50 acre/jug   
3                              243 mL                    40 acre/jug   

                                        None  
0                                       None  
1  Medium-fine to fine \ntexture (3 - 6% OM)  
2                                       None  
3                                33 acre/jug

In [62]:
# Turn tables into dataframes

dfs = []

for table in pdf.pages[512].extract_tables():
    df = pd.DataFrame(table[1:], columns=table[0])
    dfs.append(df)


It has difficulty with parsing texts in rotated diretion.


See page 510 (absolute page 514 of 704)

In the event of a text contains a lot of new line characters, do this:

In [4]:
print("".join("X\na\nM\nn \ne\ng\na\nor\nC\nn/\ne\ng\na\nor\nC".split("\n")[::-1]))

Coragen/Coragen MaX


In [65]:
dfs[0]

Products *\ns\no\nyrif\np\nor\nhl\nC  \
0   Canola (including herbicide tolerant)                        None   
1                          alfalfa looper                           C   
2                              leafhopper                               
3                            beet webworm                               
4                       Bertha \narmyworm                           C   
5                          cabbage looper                               
6                cabbage \nseedpod weevil                               
7                          clover cutworm                               
8                                 cutworm                           C   
9            diamond-back \nmoth (larvae)                           C   
10                            flea beetle                               
11                            grasshopper                           C   
12                             lygus bugs                           C   
13                    red turnip \nbeetle                               
14                               armyworm                           C   
15                          turnip aphids                               
16                   variegated \ncutworm                           C   

   X\na\nM\nn \ne\ng\na\nor\nC\nn/\ne\ng\na\nor\nC  \
0                                             None   
1                                                    
2                                                    
3                                                    
4                                                C   
5                                                C   
6                                                    
7                                                    
8                                                C   
9                                                C   
10                                                   
11                                               C   
12                                                   
13                                                   
14                                                   
15                                                   
16                                                   

   4\ne \nnt\na\nm\na\nDi\nn/\no\ng\na\nL\nn/\no\ng\ny\nC  \
0                                                None       
1                                                           
2                                                   C       
3                                                           
4                                                           
5                                                           
6                                                           
7                                                           
8                                                           
9                                                           
10                                                          
11                                                  C       
12                                                 C1       
13                                                          
14                                                          
15                                                  C       
16                                                          

   n\nhri\net\nm\na\nelt\nD\nN \nB\nF\nC/\nE\n5 \n2.\nci \ne\nol\nP\ns/\nci\ne\nD  \
0                                                None                               
1                                                                                   
2                                                                                   
3                                                   C                               
4                                                   C                               
5                                                                                   
6                                                  

In [70]:
pdf.pages[697].extract_text()

'Nozzle sizing L/acre\n694'

In [68]:
# Turn tables into dataframes

dfs = []

for table in pdf.pages[697].extract_tables():
    df = pd.DataFrame(table[1:], columns=table[0])
    dfs.append(df)


There are tables that are images. PDFPlumber won't handle tables in images. See page 694 (absolute page 698 of 704).

In [69]:
dfs[0]

IndexError: list index out of range

In [50]:
pdf.close()